### PR TITLE
Add user membership management to group management

### LINKED ISSUE #106: Add user membership management to group management
As a group owner, I would like to have buttons per user and a single button to remove or add users to my group. This should be possible on the group management page.

The "Nutzer hinzufügen" button should open a modal window that requires entering the user's email address and shows an error alert box when no user matches this email address after I submit. Otherwise, the user will be added to the group and I will be redirected to my group management page seeing that the user is now in the member list.

The small red button next to every group member in the member list should remove the desired group member from the group. After removal, the membership list should be updated for the group owner.

Add user modal
User not found while adding

### Dependencies

Depends on:

- #6 

### Acceptance Criteria

- [ ] Only the group owner is able to add or remove users from his group
- [ ] Users are members of the group after the group owner adds them to it
- [ ] Users are no longer members of a group after a group owner removes them from his group
- [ ] When a user is not found while trying to add him/her to a group, an error message is shown and the group owner can retry
- [x] Button to message group member is working (#267)
- [ ] Add user email form field should be email type
- [ ]

### PR DESCRIPTION
**Pull request by kbroschke**
closes #106

### CODE REVIEW

**Comment by KiruChaff** [comment]
We've tried implementing the feature through a form which is supposed to call the add_user method in the group controller. Unfortunately this method is not called through.

Maybe validate the resource and check if the form is correct

File: app/controllers/groups_controller.rb [hunk 1]
```diff
@@ -61,6 +61,22 @@ def leave
     end
   end
 
+
+  # PATCH /groups/1/add_user or /groups/1/add_user.json
+  def add_user
+    respond_to do |format|
+      group = @group
+      user = User.where(email: params[:user]).first
```
Comment (TobiPeterG) [inline app/controllers/groups_controller.rb:69]: should the parameter be called :email instad of :user?

**Comment by TobiPeterG** [comment]
Unfortunately, I don't have a second user account to test this feature; is there any easy way to add a new user to the database?

**Comment by DieKautz** [comment]
Whats still missing is
 * handling if the specified email belongs to no existing user and
 * shrinking the class length of the groups controller
    membership management maybe extracted to its own controller?

File: app/views/groups/edit.html.erb [hunk 2]
```diff
@@ -24,20 +24,31 @@
...
       <td><%= user.full_name %></td>
       <td><%= user.email %></td>
-      <td>
-        <a href="#" class="btn btn-sm btn-warning text-light"><i class="bi bi-chat-fill"></i></a>
-        <a href="#" class="btn btn-sm btn-primary"><i class="bi bi-person-dash-fill"></i></a>
-      </td>
+      <% if not user.admin_in?(@group) %>
+        <td>
+          <a href="#" class="btn btn-sm btn-warning text-light"><i class="bi bi-chat-fill"></i></a>
+          <% if @user.admin_in?(@group) %>
```
Comment (kbroschke) [inline app/views/groups/edit.html.erb:30]: The group controller assures that the edit page is only accessible to admins, so this will be always true.

Comment (kbroschke) [reply app/views/groups/edit.html.erb]: Should be resolved by aaf2b72

File: app/controllers/groups_controller.rb [hunk 3]
```diff
@@ -98,11 +125,20 @@ def set_group
     @group = Group.find(params[:id])
   end
 
+  def set_group_from_id
+    @group = Group.find(params[:group_id])
+  end
```
Comment (kbroschke) [inline app/controllers/groups_controller.rb:130]: Why do we need both variants?

File: app/views/groups/edit.html.erb (hunk 2 above)
Comment (kbroschke) [reply app/views/groups/edit.html.erb]: See also f4f01f2. Now CI is not failing anymore 👍🏻

**Comment by kbroschke** [comment]
I implemented the idea of moving the user management logic into a separate controller. But there is still duplicated code left, because sharing code between controllers seems to be non-trivial. I already looked into using [Concerns](https://api.rubyonrails.org/v7.0.4/classes/ActiveSupport/Concern.html). Before using a separate controller, I also experimented with [other solutions](https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/) to our problem. But in the end, just pointing the routes to the new `MembershipsController` was the easiest one to implement.

**Comment by TobiPeterG** [comment]
When a user is not found, an alert should be shown instead of a notice. I added that.
Also there was an unused function added in the GroupsController, I removed it.
Might be useful for a refactor, but not this issue here.
Otherwise, it looks good!

**Review by TobiPeterG — APPROVED** [review]
If someone could quickly look over my changes, we're good to merge!

**Comment by kbroschke** [comment]
I can't explicitly approve, but everything looks good to me. At least with refactorings tracked in #346 and in #353.
Maybe at least one additional person should review all changes (@KiruChaff or @DieKautz) because Tobias and I made so many recent changes.

**Review by DieKautz — CHANGES_REQUESTED** [review]

**Review by Encotric — APPROVED** [review]

**Review by DieKautz — APPROVED** [review]

**Pull request merged**